hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -811,55 +815,62 @@ public void reset() {
     }
 
     @Override
-    public ClientConnection getRandomConnection(boolean dataMember) {
+    public ClientConnection getRandomConnection() {
         // Try getting the connection from the load balancer, if smart routing is enabled
         if (isSmartRoutingEnabled) {
-            ClientConnection connection = getConnectionFromLoadBalancer(dataMember);
+            Member member = loadBalancer.next();
 
+            // Failed to get a member
+            ClientConnection connection = member != null ? activeConnections.get(member.getUuid()) : null;
             if (connection != null) {
                 return connection;
             }
         }
 
-        // Otherwise iterate over connections and return the very first valid
-
+        // Otherwise iterate over connections and return the first one
         for (Map.Entry<UUID, TcpClientConnection> connectionEntry : activeConnections.entrySet()) {
-            if (dataMember) {
-                UUID memberId = connectionEntry.getKey();
-
-                Member member = client.getClientClusterService().getMember(memberId);
-
-                if (member == null || member.isLiteMember()) {
-                    continue;
-                }
-            }
-
             return connectionEntry.getValue();
         }
 
         // Failed to get a connection
         return null;
     }
 
-    private ClientConnection getConnectionFromLoadBalancer(boolean dataMember) {
-        Member member;
-
-        if (dataMember) {
-            if (loadBalancer.canGetNextDataMember()) {
-                member = loadBalancer.nextDataMember();
-            } else {
-                member = null;
+    @Override
+    public ClientConnection getConnectionForSql() {
+        if (isSmartRoutingEnabled) {
+            // There might be a race - the chosen member just connected or disconnected - try a
+            // couple of times, the memberOfLargerSameVersionGroup returns a random connection,
+            // we might be lucky...
```

Comment:
I did not understand this comment. Especially the part `the chosen member just connected or disconnected`. Can you explain to me and clarify the comment if necessary?

Reply:
Consider this scenario: We get a member list, member A is there. The `memberOfLargerSameVersionGroup` method picks it. Now, A disconnects. On line 851 we'll try to get the connection to A, but it's `null`. So we try again. 
I think it can also happen that we see A in `getMemberList()` before we have a connection to it. I'm not sure about guarantees here, but the algorithm will retry as well.

Reply:
Now I see. I think `just connected or disconnected` is a little bit hard to understand. What about 

```suggestion
            // There might be a race - the chosen member might be just connected (or disconnected) - try a
            // couple of times, the memberOfLargerSameVersionGroup returns a random connection,
            // we might be lucky...
```